compiler/rustc_macros/src/symbols.rs

```diff
@@ -137,17 +142,20 @@ fn symbols_with_errors(input: TokenStream) -> (TokenStream, Vec<syn::Error>) {
     let mut keyword_stream = quote! {};
     let mut symbols_stream = quote! {};
     let mut prefill_stream = quote! {};
-    let mut counter = 0u32;
-    let mut keys =
-        HashMap::<String, Span>::with_capacity(input.keywords.len() + input.symbols.len() + 10);
+    let mut entries = HashMap::<String, Preinterned>::with_capacity(
```

Comment:
Should this use an `IndexMap<String, Span>`?

Reply:
Ooh, good idea.

Here is what that looks like: https://github.com/rust-lang/rust/commit/6eae65988a536cf2f9f02b0356383f46d9cf29eb

`indexmap::IndexMap` is not used directly elsewhere in rustc, but I was not able to use `rustc_data_structures::FxIndexMap` here because `rustc_data_structures` depends on `rustc_index` which depends on `rustc_macros` which then cannot depend on `rustc_data_structures`. This could be fixed by splitting `symbols!` out of `rustc_macros` into a different crate, or splitting `FxIndexMap` out of `rustc_data_structures`, or just redefining another `FxIndexMap` within `rustc_macros`.

https://github.com/rust-lang/rust/blob/ccb160d3432dff025ab62696a82e576bf08aa006/compiler/rustc_data_structures/src/fx.rs#L7

From looking at 6eae65988a536cf2f9f02b0356383f46d9cf29eb, I think the original code is slightly preferable to me.

`IndexMap` makes illegal states unrepresentable because it's no longer possible for there to be `Preinterned` values with duplicate `idx`, or gaps in `idx`. However, needing to handle the conversion from `usize` indices to `u32` in 4 places (instead of 1) eats into the benefit, and overall the code gets less comprehensible: for example compare `prev.span_of_name` vs `*prev.get()`.